### ISSUE TITLE
View memory layout gives incorrect layout for unions

### ISSUE THREAD

**Issue by antonilol** [rust-lang/rust-analyzer#17091]
**rust-analyzer version**: rust-analyzer version: 0.4.1830-standalone

**rustc version**: rustc 1.77.0 (aedd173a2 2024-03-17)

**editor or extension**: Code OSS (on arch linux, installed with pacman, `code 1.87.2-1`)

**relevant settings**: None

**code snippet to reproduce**: 
```rust
union Union {
    a: u16,
    b: u32,
}
```

result when hitting `Ctrl+Shift+P` and going to 'view memory layout':
![image](https://github.com/rust-lang/rust-analyzer/assets/49868160/8d713ed8-fb79-4193-9d9a-c7c616d2fd17)

when hovering on the `b: u32` field, the tooltip reports an offset of 0, although it is not placed at an offset of 0 in here:
![image](https://github.com/rust-lang/rust-analyzer/assets/49868160/068c4d9f-f5e3-4b0e-aa25-499b7413e886)

this is a cool feature, but it just needs some more polishing:

for example, it knows `PhantomData<T>` always has zero size, although in the tooltip it will say `A<{unknown}>`, but `[T; 0]` also has zero size and for that it says 'Not Available'
```rust
struct Struct<T> {
    a: usize,
    b: PhantomData<T>,
}
```

it does not account for fat pointers:
```rust
struct Struct<'a, T: ?Sized> {
    a: usize,
    b: &'a T,
}
```
this will report the `&T` being 8 bytes wide (`&{unknown}` here too). this is only correct if `T: Sized`

and for enums just size and align is displayed, nothing is done with the members

**Comment by roife** [comment]
https://github.com/rust-lang/rust-analyzer/blob/46702ffc1a02a2ac153f1d1ce619ec917af8f3a6/crates/ide/src/view_memory_layout.rs#L119-L133

Here we're simply using `.fields()` to fetch all the fields, which works for both `struct` and `union` (but does not work for `enum`). Do we need to add viewing-memory-layout for `union`s? Or simply add a check to bail out if it's a `union`. 👀

**Comment by Veykril** [comment]
Yes we need to check what kind of ADT we are dealing with here, ideally we'd render things differently for union showing that they overlap.

Tbf, this feature needs a design overhaul, it is very hard to parse imo :) Iirc Visual Studio (not Code) has a very nice looking one, and there is an extension for vscode that has a really nice one for C++ that I can't recall the name of

**Comment by roife** [comment]
> there is an extension for vscode

Might be https://marketplace.visualstudio.com/items?itemName=RamonViladomat.StructLayout. Really cool.
